### ISSUE TITLE
Make support for type comments optional

### ISSUE THREAD

**Issue by JelleZijlstra** [JelleZijlstra/typing-spec#1]
PEP 484 (and therefore this spec) imply that type checkers must support the old "type comment" syntax. However, this syntax is unnecessary since Python 3.6 and PEP 526, so there is little reason for new type checkers to support it.

I think we should no longer make support for type comments mandatory. We can keep describing the intended semantics of type comments, but say that type checkers are free to drop support for them. Code that is intended to work across type checkers (e.g., library stubs) should not use them.

This is a substantive change so it should wait for the Typing Council.

(Note that this does not include `# type: ignore` comments. Those are still useful and should continue to work.)

**Comment by dsuch** [comment]
Hello,

I was directed to this ticket through [this one](https://github.com/microsoft/pyright/issues/6526) from pyright.

This is feedback from a user who has been using type hints with comments in various places of a non-trivial product (150k lines of code) and I do not follow any Python development trackers, mailing lists, blogs and similar.

All I know, really, are PEPs and final documentation of each Python's release, and a few years ago I saw that there are two ways of providing type hints. Inline, next to a given Python object, and as comments.

I started to use type comments because I had to support Python 2.7 until very recently but, even with Python 3.8+, where I do not use type comments almost at all, they still offer an advantage in functions that have a long list of parameters, where it makes it visually more appealing to provide the types in this way:

```python

def __init__(
    self,
    *,
    pubsub,        # type: PubSub
    sub_config,    # type: anydict
    sub_key,       # type: str
    delivery_lock, # type: RLock
    delivery_list, # type: SortedList
    deliver_pubsub_msg,              # type: callable_
    confirm_pubsub_msg_delivered_cb, # type: callable_
    enqueue_initial_messages_func,   # type: callable_
    pubsub_set_to_delete,               # type: callable_
    pubsub_get_before_delivery_hook,    # type: callable_
    pubsub_invoke_before_delivery_hook, # type: callable_
) -> 'None':
```

There would be also other places where it would make my intent much clearer, like what PEP 484 envisioned here:

```python
for x in container: # type: int
   ...
```

This syntax would be useful with for loops, with blocks and the walrus, because we do not have [this one](https://discuss.python.org/t/untyped-dict-walrus-type-hints/35103) that I suggested recently:

```python
for x:'int' in container:
    ...
```

```python
for key:'str', value:'int' in untyped_dict.items():
    ...
```

```python
if status_code:'str' := untyped_dict.get('status_code', ''):
    ...
```

If you remove type comments, we will be left with this:

```python
x:'int'
for x in container:
    ...
```

A side question that I have, is that not going in the wrong direction? Does it not feel to you that it is akin to requiring people once again to declare all the variables in C upfront, before a loop begins? Similar examples could be found in C++ or Java where the direction is "becoming more like Python", which means, in this context, requiring fewer declarations in advance, before a given object can be used.


Now, that was just a side question for you to consider. I am not a core developer of Python and I will not be able to get involved in any long discussions about this subject.

The main subject, though, is that, in the future, I ask you to include information that such and such feature from a given PEP is considered provisional. I was not able to find any information about type comments being a stop-gap measure. It is good syntax, it allows me to express myself better in specific cases and these two types of syntax are not equivalent, so I do not consider comments "unnecessary".

In other words, what I see here is that the syntax of Python changes without much warning. I will cope with it and this is not a real issue but, again, I just request that if the intention in the future is to make something a temporary measure, which is perfectly fine on its own, then it would be better to have it stated explicitly in a PEP or the released documentation.

Basically, I consider it a case of miscommunication. Perhaps dropping the support for type comments at one point has been the idea all along but, not being an insider, I did not realize it because there was no information about it in the communication channels that I frequent (PEPs and documentation). Not a big problem but it is something that can be improved in the development process in the future.

Regards.

**Comment by JelleZijlstra** [comment]
Thanks for writing here. The proposal to deprecate type comment support isn't yet before the Typing Council; when that happens, there will be an opportunity to discuss it on https://discuss.python.org. Before we proceed with this deprecation, I'd want to do more research into whether type comments are still widely used. A few points, though.

First, you're probably in a very small minority when you say you prefer the type comment syntax. I have interacted with lots of people writing typed Python, and I can't remember anyone else who liked type comments. Generally, people who use them do so only because they have to support old Python versions.

It's true that in the `for x in container:` case, type comments provide a way to type the iteration variable that is arguably less awkward than the alternative. However, there is an obvious (and more type safe!) workaround: ensure that the type checker knows what type `container` has, and then you don't need to annotate `x`.

Possibly PEP 484 should have said that support could be removed when Python 2 interoperability is no longer important. However, https://peps.python.org/pep-0484/#suggested-syntax-for-python-2-7-and-straddling-code is already clear that the syntax you like for argument annotations is an optional extension meant only for Python 2 compatibility, which carries a strong implication that tools that support only Python 3 don't need to support it. Type comments for local variables (`x = 1  # type: int`) were not redundant initially, but PEP 526 (implemented in Python 3.6, about seven years ago) was clearly meant to make that syntax obsolete.

**Comment by dsuch** [comment]
Hi @JelleZijlstra - thank you for the message.

Yes, I will handle it fine, this is not a big issue from my perspective now that I do not need to support Python 2.7 and I suspect that only people migrating from 2.7 to 3.x will need to think about it in detail.

As an aside, the example with containers, unfortunately, is not that easy because at that scale of a codebase, one deals with hundreds of external dependencies, the majority of which have no type hints. That is, it is not my containers and their type is "Unknown".

This is why I [suggested](https://discuss.python.org/t/untyped-dict-walrus-type-hints/35103) that this syntax be added to the language:

```python
for key:'str', value:'int' in untyped_dict.items():
    ...
```

After all, we have this:

```python
x:'int' = get_object()
```
Then, how is it different from this one, that we do not have?

```python
if x:'int' := get_object():
    ...
```

To me, this looks like a simple oversight. But this is a side discussion, I assume that you are a core Python developer and, if you like this proposal, and if you can somehow make it a proposal that other core developers consider, that will be great, because this is a real-world use-case, based on an actual need, that would definitely help with my development work, and I put it in here because you mentioned containers.

Thank you.

**Comment by JelleZijlstra** [comment]
python/typing#1533
